src/attached_probe.cpp

```diff
@@ -172,55 +181,12 @@ AttachedProbe::AttachedProbe(Probe &probe,
                              std::optional<int> pid,
                              BPFtrace &bpftrace,
                              bool safe_mode)
-    : probe_(probe), progfd_(prog.fd()), bpftrace_(bpftrace)
+    : probe_(probe),
+      progfd_(prog.fd()),
+      bpftrace_(bpftrace),
+      pid_(pid),
+      safe_mode_(safe_mode)
```

Comment:
Why allow construction of `AttachedProbe` that hasn't been attached? IOW this is partial construction, as destructor depends on full construction. See `detach_iter()`, `detach_raw_tracepoint()`, etc.

It'd be more RAII friendly to have a static factory method that returns a fully attached `AttachedProbe`, as the name would imply. ie.

```
class AttachedProbe {
public:
  static Result<AttachedProbe> make_attached_probe(Probe &probe, ...);
  ~AttachedProbe();
  ...
private:
  AttachedProbe(Probe &probe, ...);
  ...
};
```

This removes the need for everything in the AttachedProbe destructor to consider whether or not attachment has been done. It also saves the `ap->attach()` calls from the various call sites. It also avoids the semantic drift in naming (see below).

Reply:
> This removes the need for everything in the AttachedProbe destructor to consider whether or not attachment has been done.

What conditional logic in the destructor are you referring to? The closing of potentially negative or un-initialized file descriptors?

Reply:
Right, we have those right now. But in future if detachment logic is more complex we don't want to consider attached and unattached cases.

Reply:
@danobi and I discussed offline - going to save the refactor for a separate PR.